services/victoriametrics/vmalert.go

```diff
@@ -0,0 +1,204 @@
+)
+
+// VMAlert is responsible for interactions with victoria metrics.
+type VMAlert struct {
+	enabled             bool
+	baseURL             *url.URL
+	client              *http.Client
+	alertingRules       *prometheus.AlertingRules
+	cachedAlertingRules string
+
+	l    *logrus.Entry
+	sema chan struct{}
+}
+
+// NewVMAlert creates new Victoria Metrics Alert service.
+func NewVMAlert(alertRules *prometheus.AlertingRules, baseURL string, params *models.VictoriaMetricsParams) (*VMAlert, error) {
+	if !params.Enabled {
+		return &VMAlert{}, nil
+	}
+	u, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	return &VMAlert{
+		enabled:       params.Enabled,
+		alertingRules: alertRules,
+		baseURL:       u,
+		client:        new(http.Client),
+		l:             logrus.WithField("component", "vmalert"),
+		sema:          make(chan struct{}, 1),
+	}, nil
+}
+
+// Run runs VMAlert configuration update loop until ctx is canceled.
+func (svc *VMAlert) Run(ctx context.Context) {
+	if !svc.enabled {
+		return
+	}
+	svc.l.Info("Starting...")
+	defer svc.l.Info("Done.")
+	alertingRules, err := svc.alertingRules.ReadRules()
+	if err != nil {
+		svc.l.Warnf("Cannot load alerting rules: %s", err)
+	}
+	svc.cachedAlertingRules = alertingRules
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-svc.sema:
+			// batch several update requests together by delaying the first one
+			sleepCtx, sleepCancel := context.WithTimeout(ctx, updateBatchDelay)
+			<-sleepCtx.Done()
+			sleepCancel()
+
+			if ctx.Err() != nil {
+				return
+			}
+
+			if err := svc.updateConfiguration(ctx); err != nil {
+				svc.l.Errorf("Failed to update configuration, will retry: %+v.", err)
+				svc.RequestConfigurationUpdate()
+			}
+		}
+	}
+}
+
+// RequestConfigurationUpdate requests VMAlert configuration update.
+func (svc *VMAlert) RequestConfigurationUpdate() {
+	if !svc.enabled {
+		return
+	}
+	select {
+	case svc.sema <- struct{}{}:
+		ctx, cancel := context.WithTimeout(context.Background(), configurationUpdateTimeout)
+		defer cancel()
+		err := svc.updateConfiguration(ctx)
+		if err != nil {
+			svc.l.WithError(err).Errorf("cannot reload configuration")
+		}
+	default:
+	}
+}
+
+// IsReady verifies that VMAlert works.
+func (svc *VMAlert) IsReady(ctx context.Context) error {
+	if !svc.enabled {
+		return nil
+	}
+	// check VMAlert /health API
+	u := *svc.baseURL
+	u.Path = path.Join(u.Path, "health")
+	req, err := http.NewRequestWithContext(ctx, "GET", u.String(), nil)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	resp, err := svc.client.Do(req)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer resp.Body.Close() //nolint:errcheck
+	b, err := ioutil.ReadAll(resp.Body)
+	svc.l.Debugf("VMAlert: %s", b)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != 200 {
+		return errors.Errorf("expected 200, got %d", resp.StatusCode)
+	}
+
+	svc.l.Debugf("%s", b)
+
+	return nil
+}
```

Comment:
Could it be a shared function in Prometheus package?

Reply:
Same as above, Prometheus package is going to be removed.